app/src/androidMain/kotlin/de/westnordost/streetcomplete/ui/common/LengthInput.kt

```diff
@@ -0,0 +1,243 @@
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.times
+import de.westnordost.streetcomplete.data.meta.LengthUnit
+import de.westnordost.streetcomplete.osm.Length
+import de.westnordost.streetcomplete.ui.util.onlyDecimalDigits
+import de.westnordost.streetcomplete.ui.util.validFeetInput
+import de.westnordost.streetcomplete.ui.util.validInchInput
+import kotlin.math.pow
+
+enum class FootInchAppearance { PRIME, UPPERCASE_ABBREVIATION }
+
+@Composable
+fun LengthInput(
+    selectedUnit: LengthUnit,
+    footInchAppearance: FootInchAppearance,
+    currentLength: Length?,
+    syncLength: Boolean,
+    onLengthChanged: (Length?) -> Unit,
+    maxFeetDigits: Int,
+    maxMeterDigits: Pair<Int, Int>,
+    modifier: Modifier = Modifier,
+    colors: TextFieldColors = TextFieldDefaults.textFieldColors(),
+) {
+    if (selectedUnit == LengthUnit.METER) {
+        LengthInputMeters(
+            currentLength,
+            syncLength,
+            onLengthChanged,
+            maxMeterDigits,
+            modifier,
+            colors,
+        )
+    } else if (selectedUnit == LengthUnit.FOOT_AND_INCH) {
+        LengthInputFootInch(
+            currentLength,
+            footInchAppearance,
+            syncLength,
+            onLengthChanged,
+            maxFeetDigits,
+            modifier,
+            colors,
+        )
+    }
+}
+
+@Composable
+private fun LengthInputMeters(
+    currentLength: Length?,
+    syncLength: Boolean,
+    onLengthChanged: (Length?) -> Unit,
+    maxMeterDigits: Pair<Int, Int>,
+    modifier: Modifier,
+    colors: TextFieldColors,
+) {
+    var lengthMeterText by remember { mutableStateOf("") }
+
+    LaunchedEffect(currentLength, syncLength) {
+        // Sync from the length only when explicitly allowed (e.g. when inserting data from AR measurement)
+        if (syncLength && currentLength is Length.Meters) {
+            // Clamp meters to maximum allowed numbers
+            val upper = 10.0.pow(maxMeterDigits.first) - 10.0.pow(-maxMeterDigits.second)
+            val clamped = currentLength.meters.coerceIn(0.0..upper)
+            lengthMeterText = String.format("%.2f", clamped)
+
+            // For the unlikely case, that clamping has happened: Propagate change back to parent
+            // so that when still published, the displayed value is used.
+            if (clamped != currentLength.meters) {
+                onLengthChanged(Length.Meters(clamped))
+            }
+        }
+    }
+
+    TextField(
+        value = lengthMeterText,
+        onValueChange = { input: String ->
+            if (input.isEmpty()) {
+                onLengthChanged(null)
+                lengthMeterText = input
+            } else if (onlyDecimalDigits(
+                    input,
+                    maxMeterDigits.first,
+                    maxMeterDigits.second
+                )
+            ) {
+                onLengthChanged(Length.Meters(input.toDouble()))
```

Comment:
So this is going to fail for `"3,5"`, AFAIK. We should test what happens on e.g. a German locale and one inputs a decimal with a comma. If it fails, we need some generic solution that we can apply to all the other text fields that allow decimal input.
If there exists nothing in the Kotlin standard library itself, we might want to define an own extension function on `String` - `anyDecimalToDouble` or something. 🤷

(See also https://en.wikipedia.org/wiki/Decimal_separator )